src/dotnet/MLSearch.Service/Indexing/ChatIndexerWorker.cs

```diff
@@ -1,29 +1,69 @@
-using ActualChat.MLSearch.ApiAdapters;
-
 
 namespace ActualChat.MLSearch.Indexing;
 
 
 internal class ChatIndexerWorker(
     IDataIndexer<ChatId> dataIndexer,
-    ILoggerSource loggerSource
+    ILogger<ChatIndexerWorker> log
 ) : IChatIndexerWorker
 {
+    private record class Job(Task Completion, CancellationTokenSource Cancellation);
     private const int ChannelCapacity = 10;
-    private ILogger? _log;
-    private ILogger Log => _log ??= loggerSource.GetLogger(GetType());
 
     private readonly Channel<MLSearch_TriggerChatIndexing> _channel =
-        Channel.CreateBounded<MLSearch_TriggerChatIndexing>(new BoundedChannelOptions(ChannelCapacity){
+        Channel.CreateBounded<MLSearch_TriggerChatIndexing>(new BoundedChannelOptions(ChannelCapacity) {
             FullMode = BoundedChannelFullMode.Wait,
-            SingleReader = false,
+            SingleReader = true,
```

Comment:
I do believe this is a negative change, here is the reason.
With this change you're trying to put all triggered chats to index. This means in case of slow indexing (and it's actually a normal behavior) those chat identifiers will be buffered in the app memory vs in a queueing service. However the later is built to rescale at runtime, and, what is more important, it's persistent. With this change it will soak all the load into the memory of the app and in case of a failure all of those triggered events would disappear. Not only a single one, but all. Having a channel with a limited size buffer was already a compromise.

Reply:
I introduced single dispatch flow to handle incoming events in this channel. This is why I switched to single reader.